src/System/Interactions.jl

```diff
@@ -372,6 +373,107 @@ function set_forces_aux!(B, dipoles::Array{Vec3, 4}, ints::Interactions, i::Int,
     end
 end
 
+function set_forces_aux_homog!(B, dipoles::Array{Vec3, 4}, buf::Array{Vec3,4}, ints::Interactions, i::Int, sys::System{N}) where N
+    (; latsize) = sys
+
+    cells = all_cells(sys);
+
+    subLatBuffer = view(buf,:,:,:,1);
+
+    # Single-ion anisotropy only contributes in dipole mode. In SU(N) mode, the
+    # anisotropy matrix will be incorporated directly into ℌ.
+    if N == 0
+
+        for cell in cells
+            s = dipoles[cell, i]
+            B[cell, i] -= energy_and_gradient_for_classical_anisotropy(s, ints.aniso.stvexp)[2]
+        end
+        #f = s -> energy_and_gradient_for_classical_anisotropy(s,ints.aniso.stvexp)[2]
+        #@. xx = f(dipoles)
+        #@. B -= xx
+    end
+
+
+    # Heisenberg exchange
+    for (; isculled, bond, J) in ints.heisen
+        isculled && break
+        # Forwards interaction: j there contributes to i here
+        fastcircshift3!(subLatBuffer,view(dipoles,:,:,:,bond.j),-bond.n)
+        subLatBuffer .*= J
+        @views B[:,:,:,bond.i] .-= subLatBuffer
+
+        # Backwards interaction: i here contributes to j there
+        fastcircshift3!(subLatBuffer,view(dipoles,:,:,:,bond.i),+bond.n)
+        subLatBuffer .*= J'
+        @views B[:,:,:,bond.j] .-= subLatBuffer
+    end
+    # Quadratic exchange matrix
+    for (; isculled, bond, J) in ints.exchange
+        isculled && break
+        for cellᵢ in cells
+            cellⱼ = offsetc(cellᵢ, bond.n, latsize)
+            sᵢ = dipoles[cellᵢ, bond.i]
+            sⱼ = dipoles[cellⱼ, bond.j]
+            B[cellᵢ, bond.i] -= J  * sⱼ
+            B[cellⱼ, bond.j] -= J' * sᵢ
+        end
+    end
+    # Scalar biquadratic exchange
+    for (; isculled, bond, J) in ints.biquad
+        isculled && break
+        for cellᵢ in cells
+            cellⱼ = offsetc(cellᵢ, bond.n, latsize)
+            sᵢ = dipoles[cellᵢ, bond.i]
+            sⱼ = dipoles[cellⱼ, bond.j]
+
+            if sys.mode == :dipole
+                Sᵢ = (sys.Ns[cellᵢ, bond.i]-1)/2
+                Sⱼ = (sys.Ns[cellⱼ, bond.j]-1)/2
+                S = √(Sᵢ*Sⱼ)
+                # Renormalization introduces a factor r and a Heisenberg term
+                r = (1 - 1/S + 1/4S^2)
+                B[cellᵢ, bond.i] -= J * (2r*sⱼ*(sᵢ⋅sⱼ) - sⱼ/2)
+                B[cellⱼ, bond.j] -= J * (2r*sᵢ*(sᵢ⋅sⱼ) - sᵢ/2)
+            elseif sys.mode == :large_S
+                B[cellᵢ, bond.i] -= J * 2sⱼ*(sᵢ⋅sⱼ)
+                B[cellⱼ, bond.j] -= J * 2sᵢ*(sᵢ⋅sⱼ)
+            elseif sys.mode == :SUN
+                error("Biquadratic currently unsupported in SU(N) mode.")
+            end
+        end
+    end
+end
+
+function fastcircshift3!(dest,src,shift)
```

Comment:
I am curious what happens if you try Julia's built-in `circshift!` -- does it not work well for arrays of 3-vectors?

Reply:
Built-in `circshift!` is generic in the number of indices. For vectors and matrices it is non-allocating, but it is allocating (likely because it doesn't inline aggressively enough?) for rank 3 tensors and higher. `fastcircshift3!` is just exactly built-in `circshift!` specialized to rank 3 tensors and unrolled explicitly.